### DISCUSSION TITLE
Question: global way to always print to stdout/stderr?

### DISCUSSION THREAD

**Discussion by vsoch** [nextflow-io/nextflow#3421] [Q&A] [answered]
Apologies if this is a noob question! I am testing the machine learning workflow: https://github.com/nextflow-io/ml-hyperopt/
and have gotten it to be able to run with the Flux Executor - as expected, nextflow kicks off a bunch of jobs!

![image](https://user-images.githubusercontent.com/814322/203217726-89c85f56-3394-4655-ae47-5509b57e8643.png)

Of course then I quickly realized they don't have terminal output (e.g., the flux process can show this when you click on the job) and first I thought it was because I specified `--output` for the FluxExecutor, but now I'm realizing that nextflow can handle different kinds of output, and it's usually defined in the progress (e.g., I found an example here https://gitter.im/nextflow-io/nextflow?at=5afa7ecfd245fe2eb7af6dc5 and the docs here https://www.nextflow.io/docs/edge/process.html#outputs). It looks like the specification for where to write output, in the case of this workflow, happens in the modules: https://github.com/nextflow-io/ml-hyperopt/tree/master/modules. I also found "debug" and I think that is of the same thread - it needs to be defined in each workflow file. Given something that already exists, I've have to custom write each of the files to better control the output.

So my question - which I hope there is a stupidly simple answer to - is if there is a way to globally say "Always print your texty stuffs to stdout/stderr" and that way the Flux RESTFul API interface can show it easily. If there is a way to control it within the FluxExecutor (and I'm able to add some special flag) that would be great to know too - basically I just want stuff to print to the terminal within the process that is running the job. Thank you!

**Comment by mribeirodantas** [comment, marked as answer]
Instead of adding `debug true` to the beginning of every process block, you can set it once in the `nextflow.config` file. Simply adding the line below to the `nextflow.config` file in your launching directory...
```
process.debug = true
```

will make the following Nextflow script print hello and then world to the standard output.
```
process A {

  output:
    stdout
  script:
  """
  echo 'hello'
  """
}

process B {

  output:
    stdout
  script:
  """
  echo 'world'
  """
}

workflow {
  A()
  B()
}
```

Try removing the line from the `nextflow.config` file and you'll see the hello and world will no longer be printed to the standard output.

There is also another way to do that, without having to create/change the `nextflow.config` file. Simply add `-process.echo` to the command line. For example:
```
➜  ~ nextflow run ex2.nf -process.echo
```

One last trick! Even if Nextflow doesn't print to your standard output, it must print somewhere, right? You can see the output by checking the content of the `.command.out` file that is located within every task folder (try typing `tree work` where you launched your pipeline to see the folders). If you have just run the pipeline I wrote above without seeing the hello and world, you can see them by typing the following command:
```
➜  ~ nextflow log last | while read line ; do cat $line/.command.out ; done
```

**Reply by vsoch** [reply]
Perfect! I'll test this now (you are so speedy!) for context I'm testing with https://github.com/rse-ops/flux-hpc/tree/main/nextflow/ml-hyperopt and as soon as we have that output we'll have logs in the user interface.

Testing now, will report back.

**Reply by vsoch** [reply]
@mribeirodantas that was close but not quite what I needed! So with debug I am able to see the output in the stdout of the terminal where I ran the nextflow command. However, using Flux as an executor (meaning each step is submit with a flux job) I'd expect to see the output in the job for flux (which I do not!) Perhaps is the issue that it's still going to https://github.com/nextflow-io/nextflow/blob/aa974d445217ebf0f30d98dcdbdbbfb0066c4082/modules/nextflow/src/main/groovy/nextflow/executor/FluxExecutor.groovy#L64?

The way flux works when you submit to it, given that there is stdout or stderr you can attach to the job and see the output, and that's what shows up in the UI I've developed. With debug = true I don't see anything, so I think the debug is just having it all printed in the top level calling process, if that makes sense.

Thanks for your help!

**Reply by mribeirodantas** [reply]
Yes, [`debug`](https://www.nextflow.io/docs/latest/process.html?highlight=debug#debug) set to true forwards the process stdout to the current top running process stdout file, showing it in the shell terminal. If you don't use `-process.echo` or the `debug` process directive, it will be only redirected to the `.command.out` (or `.command.err` for standard error) file(s) of the respective task.

I don't think there's a way to prevent this capture 😞

Do you know of anything, @bentsherman?

**Reply by vsoch** [reply]
Ideally I'd like to be able to have the stdout and stderr keep it's own output, if that makes sense! And if there is a way, even within an executor, I'm happy to try implementing it - I just added support for Flux this weekend and happy to do more tweaks! 

As an alternative, is there a global way to define an additional step for all the processes? E.g., something that minimally cats the `.command.out`? There must be some hack we can figure out!

Thank you for your help with this!

**Reply by mribeirodantas** [reply]
Maybe automatically adding a new line in the .command.sh to cat the .command.out/err file? 😅 I'm not sure about the best way to implement this.

**Reply by vsoch** [reply]
Woke up with an idea - I think I just got it working! :partying_face: 

![image](https://user-images.githubusercontent.com/814322/203309589-d6e7af1e-3681-4a34-a5cd-fddf9c30641b.png)

**Reply by vsoch** [reply]
Workaround for FluxExecutor here: https://github.com/nextflow-io/nextflow/pull/3424.

**Reply by mribeirodantas** [reply]
Oh, very nice to hear that you managed to make it work 🥳 . Congrats!

**Reply by bentsherman** [reply]
Hi @vsoch , I scanned this thread and I'm still not sure what you're trying to do. Since your bugfix PR is unlikely to be merged as is, I want to see if you can accomplish what you want with existing features.

There are a bunch of ways to capture command outputs from a Nextflow task:
- `.command.out` and `.command.err` are the stdout and stderr of the task script `.command.sh`
- `.command.log` is the combined stdout/stderr of `.command.run`, the wrapper script that calls `.command.sh` (it will include the contents of the .out and .err files as well as any output from the wrapper script)
- if you specify a `stdout` process output, it will capture the stdout of `.command.sh` into a channel... I think it still outputs to `.command.out` but I'm not sure
- if you specify `process.debug = true`, the task stdout will also be sent to the Nextflow stdout, in addition to `.command.out`

So if you want to display the task output in a web interface, you could just read the `.command.out` or `.command.log` from the task directory, right?

**Reply by vsoch** [reply]
> .command.out and .command.err are the stdout and stderr of the task script .command.sh

Yes, and we originally were writing all output to these logs and didn't see anything in the terminal.

>  .command.log is the combined stdout/stderr of .command.run, the wrapper script that calls .command.sh (it will include the contents of the .out and .err files as well as any output from the wrapper script)

Ditto, I'm not interested in writing to a file.

> if you specify a stdout process output, it will capture the stdout of .command.sh into a channel... I think it still outputs to .command.out but I'm not sure

I tried adding this to the main main.nf after reading about it in the docs and got an error that the syntax was invalid. 

>  if you specify process.debug = true, the task stdout will also be sent to the Nextflow stdout, in addition to .command.out

This actually didn't work as you'd expect - it printed logs / output to the _main_ process (for me, web server logs that nobody will see) but the worker stdout/stderr was still empty (what Flux can return to the UI for the user).

> So if you want to display the task output in a web interface, you could just read the .command.out or .command.log from the task directory, right?

No, we need the stdout/stderr to be in the terminal output of the process. We can't do a custom file reading for a specific workflow manager - Flux is treating it as just any command, and it will produce output.

**Reply by bentsherman** [reply]
I think I understand. Flux needs to consume the task output through stdout/stderr, but Nextflow needs to consume the stdout/stderr through files.

But when you submit a task to Flux with `--output=/path/to/.command.log`, then Flux knows where the process output will be. So why not have Flux simply read from that file? Similarly, if `--output` isn't specified, Flux should read the stdout/stderr of the process, which sounds like what it does now. That is a workflow-agnostic solution.

The only other path I see is to do something with `tee` to pipe the output of `.command.run` to both the terminal and `.command.log`.

**Reply by vsoch** [reply]
Yes, Flux does know that, but the Flux Operator (running a generic command) cannot be implemented to require reading in a specific file for the server for a specific workflow manager. That's a no OP. If I were building a specific web interface for nextflow that would be fine, but it's not for that.

> The only other path I see is to do something with tee to pipe the output of .command.run to both the terminal and .command.log

The PR I've opened is the solution that I need, and it just adds an additional config value for the FluxOperator. So I'm not sure why you see this as "the only other path" - I've found a path that works perfectly, implemented it, and did a PR to update the executor.

**Reply by pditommaso** [reply]
There's something I'm not understanding. The presence *or* the absence of the `--output=/path/to/.command.log` option should not change how nextflow prints the tasks output. Does it? if so how?

**Reply by bentsherman** [reply]
I'm sorry, I'm not familiar with the Flux Operator or it's constraints. Is there any documentation for it? Or can you elaborate on what it does?

As it stands, I just don't think it's a Nextflow-specific issue. Processes can send their output to stdout/stderr or to a file, so it's generally good to be able to send and receive output both ways, in any context. The problem with the PR is that by sending the output to the terminal you're disabling the `.command.log` file, which Nextflow users generally expect to have in order to debug issues after the fact, regardless of the executor they use.

Ultimately, Flux is in control of what to do with process output. Nextflow just provides the command and the output path... Flux could execute the command in a way that sends the output to both stdout and a file. Right?

**Reply by vsoch** [reply]
> The presence or the absence of the --output=/path/to/.command.log option should not change how nextflow prints the tasks output. Does it? if so how?

It doesn't influence how Nextflow prints anything - either way, I see the same terminal output in the main process. However, in the (worker?) process, changing --output determines if it's printed to stdout/stderr. Given that `--output` is not defined, we get the printing in the process for Flux to find.

I think this is really simple. I need a flag that says "Don't define the `--output` variable" and just run without it. Right now the default in the executor is always to define it, and I need a flag that allows me to skip that. Please let me know the best way to implement this.

**Reply by vsoch** [reply]
> Ultimately, Flux is in control of what to do with process output. Nextflow just provides the command and the output path... Flux could execute the command in a way that sends the output to both stdout and a file. Right?

Yes, and that's why I added the config value to control how flux does this. I feel like we are talking around one another.

**Reply by vsoch** [reply]
This is what I need. https://github.com/nextflow-io/nextflow/blob/b6958c70ccd8ed03dec89cba57e2cddcfb68c8e2/modules/nextflow/src/main/groovy/nextflow/executor/FluxExecutor.groovy#L67-L69

**Reply by bentsherman** [reply]
I understand that this snippet is a simple solution to your problem, but it also causes the `.command.log` file to not be written, and this file is a standard file that many users and external tools use to e.g. debug errors. That may be fine if they're using your Flux-based UI, but if they're running Nextflow+Flux some other way then they'll be out of luck.

What I'm proposing is to change Flux itself:
- if `--output` is not specified, Flux should consume the stdout/stderr (as it currently does)
- if `--output` is specified, Flux should consume the output file

It's a WM-agnostic solution, and it will probably be beneficial for Flux in the long run, because you're gonna run into this same problem with any tool that redirects process outputs to files.

@pditommaso if you're fine with having a config option that disables the `.command.log` file in order to make @vsoch 's use case work, we can clean up the PR and merge it. But I think this issue would be better addressed in Flux rather than Nextflow.

**Reply by vsoch** [reply]
Flux already is doing what you suggest - if I specify output, I get it written to a file. If I don't, it's streamed into the terminal process. It's just in the FluxExecutor within nextflow that I need to be able to control this - right now it's _always_ written, an oversight on my part (but mimicking the other executors). This is a different (possibly new) use case because most HPC executors don't come with interfaces that one could see a specific job running. So - I'm suggesting either addition a new config for the FluxOperator (the PR) or if that isn't supported, I could do a hack to add a "special" flag in the extra command options, loop through that list looking for it, and then set (or don't set `--output`) depending on that.

**Reply by bentsherman** [reply]
Right, but then you have this web app that is forwarding the output from the stdout to the frontend, right? That's the part I'm talking about -- the web app could consume stdout or a file based on the flux submit command. I can't tell if the web app is part of Flux or an external service.

**Reply by vsoch** [reply]
No it's actually not - when you attach to the process with Flux, it sees the output. So regardless of the web app (or not) if we ran this with Nextflow and flux right now, when the user did a `flux job attach <jobid>` to get to the process, they would see nothing. It would be confusing at best.

**Reply by vsoch** [reply]
The RESTFul API "web app" I've been showing uses a Python binding to flux, and just does basic commands you could also run on the command line with flux to list jobs, submit jobs, and get output (that attach command). It's just a thin wrapper. But if the thing that Flux is running (or Nextflow submitting jobs) is essentially hiding the output from the process, the user experience is going to be bad. So we need to fix this, regardless of the web app I'm building or not I think.

**Reply by vsoch** [reply]
So here are my two suggestions:

- allow a FluxExecutor flag to control adding --output for flux submit (or not) - perhaps other executors could find this useful too.
- refactor the current PR to have a special extra flag in the extra submit params that are allowed. It's a bit of a hack and ugly to loop through and look through the list, but if you are strongly against the first suggestion, if this is what I need to get it working it works for me. I'll just want to very clearly document how to do it for the user.

**Reply by bentsherman** [reply]
Okay, so then we're back to my first suggestion -- Flux should be able to forward process output from stdout or a file based on the presence of `--output`. So when a user runs e.g. `flux job attach <jobid>`, Flux should either forward the terminal output or cat the output file given by `--output`. Alternatively, if `--output` is given then Flux should launch the command with `tee` so that the command output goes to both the terminal and the output file.

Flux has complete control over how to execute the command, i.e. it has to do the equivalent of `command > output.txt`, so it should be able to pipe the process output however it wants.

**Reply by vsoch** [reply]
@grondo is this possible with Flux? I didn't think so - it seems like an either/or thing.

I'm not sure why there is so much pushback to essentially adding an optional flag - I'm going to refactor the PR to allow this as a custom parameter (we can pretend for Flux) but it will be parsed specially. 

I think given this is specifically the _Flux Executor_ you should allow us to customize it it within NextFlow _in the way that best supports using Flux with Nextflow_. Asserting that the underlying workflow manager needs to change is a bit much. Would you go insist changes to the slurm developers for a tiny thing like this? I don't quite understand the push back - I am trying to make it easy to use Flux with Nextflow but you are fighting against me.

**Reply by vsoch** [reply]
okay I've come up with a compromise to allow the boolean flag to go in as a clusterOption, no additional config arguments for nextflow, and we get the same functionality (I've rebuilt my example from scratch and confirmed it works) and I've also added a test to confirm `--output` is not given to Flux when it's provided. That's the flag that Flux uses to determine to write to file or stdout, and we simply need to be able to specify it (or not) here. To be clear, this functionality already exists in Flux, and we need it exposed to the user here.

https://github.com/nextflow-io/nextflow/blob/1a0486438a42304ce451e65ccf8473e43466db52/modules/nextflow/src/main/groovy/nextflow/executor/FluxExecutor.groovy#L65-L82.

**Reply by grondo** [reply]
> @grondo is this possible with Flux? I didn't think so - it seems like an either/or thing.

I'm not sure I'm following this entire thread, but from a Flux perspective:

> Flux should be able to forward process output from stdout or a file based on the presence of --output. So when a user runs e.g. flux job attach <jobid>, Flux should either forward the terminal output or cat the output file given by --output

This might not be a good idea to the real `flux job attach` command, since there is no way to guarantee that more than one job didn't specify the same `--output` file and Flux is designed to run many thousands of jobs with uptimes of months or more.

However, @vsoch, would it work to have a python script or function that could grab output from the Flux output eventlog, but would detect if the output was redirected to a file and would try to read from that file instead? I could help write a prototype of that function. It might come in handy if there are other workflow systems using Flux which require every job output in a file.

Otherwise, perhaps "wrapping" the nextflow jobs submitted to Flux with a `tee` seemed like a workable solution for this case. Actually, I wonder if we could add a mode to the job shell that would redirect to a file _and_ the kvs output eventlog. Feel free to open an issue if this would be a nice feature.

**Reply by vsoch** [reply]
At the end of the day, we'd need a way for the user to request this functionality from the Nextflow config, either in the config or one off. I'm not sure how a function like that would help if we aren't allowed to do that. This is also a web server - we aren't really able to run commands like tee willy nilly to get logs - the way we get logs consistently across any command launched is with the flux jobs attach. I can't be implementing custom logic just for Nextflow.

**Reply by pditommaso** [reply]
I've replied in the PR https://github.com/nextflow-io/nextflow/pull/3424#discussion_r1030162206

**Reply by bentsherman** [reply]
Thanks @grondo for chiming in.

> Otherwise, perhaps "wrapping" the nextflow jobs submitted to Flux with a tee seemed like a workable solution for this case. Actually, I wonder if we could add a mode to the job shell that would redirect to a file and the kvs output eventlog. Feel free to open an issue if this would be a nice feature.

It took me a while to grok the whole problem here, but this is the solution I settled on at the end of the day. Flux should just always send the process output to stdout, but then also to a file if `--output` is specified. I know PBS works this way, possibly SLURM as well. I suspect you will eventually have more users who want the terminal output and the log file, not just for Nextflow but any other tool that works this way.

Anyway, for now I think we will just merge @vsoch 's PR since it is an opt-in feature.

**Reply by bentsherman** [reply]
@vsoch I apologize for starting out by saying your PR probably wouldn't be merged, that was rash of me 😄

**Reply by grondo** [reply]
>  I know PBS works this way, possibly SLURM as well. I suspect you will eventually have more users who want the terminal output and the log file, not just for Nextflow but any other tool that works this way.

Great advice! Though I'm pretty sure Slurm can't do this (former Slurm developer here). We should be able to support this mode though, and I can see it will definitely be useful. :+1:
